### ISSUE TITLE
TopP op

### ISSUE THREAD

**Issue by yieldthought** [tenstorrent/tt-metal#6194]
During the decoding loop the standard is to use a combination of top-k and top-p sampling to select the token from the output logits. For performance reasons we can't afford to be doing this on the host and leaving the device idle during that and the 2 additional transfers.

Open to removing the P1_feature_needed label - this is needed _only_ for performance. We can demonstrate functionality without it. However, we cannot really ship performance models to customers without it.

This affects all LLMs.

**Comment by tarafdarTT** [comment]
@yieldthought , is this a superset of what we are working on for mixtral.
@sraizada-tt does indexed fill, embeddings and non-zero indices cover this?

**Comment by sraizada-tt** [comment]
Yes, we are doing this with argmax, then indexed fill (to mask out the 1st max value) and then a second argmax. At some, point if we could get an integrated top-2 op that would be great, but we have a workaround for now.

**Comment by sraizada-tt** [comment]
No, indexed fill, embeddings and non-zero won't be enough. We can do upto top-2 using argmax, masking, argmax but anything over that would be too low on performance

**Comment by tarafdarTT** [comment]
@sraizada-tt sorry if this is an ignorant question, would we need this and topk (that's being handled in another issue)https://github.com/tenstorrent-metal/tt-metal/issues/5906

**Comment by yieldthought** [comment]
Yes, it’s very common in LLM decoding to use both top-k and top-p sampling and we need to support both.

**Comment by mtairum** [comment]
Are there any updates on this issue?

**Comment by jliangTT** [comment]
It is critical for all the LLM, but it is at the end of the model.

**Comment by jliangTT** [comment]
c.c. @ttmtrajkovic

**Comment by uaydonat** [comment]
The latest update is that mistral/mixtral is able to work-around topp using argmax which does not seem to affect the results.

Falcon/llama uses topk_topp for token generation but the measurements show that this is not taking any significant time when it is run on the host. For now, we can lower the priority of topp.

**Comment by jliangTT** [comment]
Will work on this after TopK (current WIP). TopP will be based on TopK operator

**Comment by uaydonat** [comment]
@mtairum @sraizada-tt are we done with topp? Are we using on-device topp?

**Comment by sraizada-tt** [comment]
done : https://github.com/tenstorrent/tt-metal/pull/17136
